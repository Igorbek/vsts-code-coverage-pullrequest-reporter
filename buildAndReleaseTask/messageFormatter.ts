import { CodeCoverageData } from "azure-devops-node-api/interfaces/TestInterfaces";

interface CodeCoverageColumn {
  label: string;
  stat: CodeCoverageData;
}

/*
Sample format
|  | master | % | PR | % |
|-:|-:|-----------:|-----------:|--:|
| **Lines** |
| covered | 800 |  84.7% | (+29 :arrow_up_small:) 829 | (+.5%) 85.2% |
| uncovered | 144 |  15.3% | 144 | (-.5%) 14.8% |
| total | 944 | | (+29) 973 |
| **Branches** |
| covered | 800 |  84.7% | (+29) 829 | 85.2% :arrow_up_small: .5%|
| uncovered | 144 |  15.3% | 144 | -.5% :arrow_down_small: 14.8% |
| total | 944 | | (+29) 973 |
 */

function formatNumber(n: number) { return Number(n.toFixed(1)).toString(); }
function getPercentage(n: number, total: number) { return total == 0 ? 0 : (n * 100 / total); }
function formatPercentage(n: number, total: number) { return formatNumber(getPercentage(n, total)) + '%'; }
function getCheck(check: number, change: number) {
  return check === 0 ? '' :
    check * change < 0 ? ' :warning:' :
    ' :white_check_mark:';
}
function formatNumberChange(newValue: number, oldValue: number, checks: number = 0, numberDecorator: string = '', sign: string = '') {
  return newValue == oldValue ? `${numberDecorator}${formatNumber(newValue)}${sign}${numberDecorator}${getCheck(checks, 0)}`:
    newValue > oldValue ? `(+${numberDecorator}${formatNumber(newValue - oldValue)}${sign}${numberDecorator} :arrow_up_small:) ${numberDecorator}${formatNumber(newValue)}${sign}${numberDecorator}${getCheck(checks, 1)}` :
    `(-${numberDecorator}${formatNumber(oldValue - newValue)}${sign}${numberDecorator} :arrow_down_small:) ${numberDecorator}${formatNumber(newValue)}${sign}${numberDecorator}${getCheck(checks, -1)}`;
}
function formatPercentageChange(newValue: number, newTotal: number, oldValue: number, oldTotal: number, checks: number = 0, numberDecorator: string = '') {
  return formatNumberChange(getPercentage(newValue, newTotal), getPercentage(oldValue, oldTotal), checks, numberDecorator, '%');
}

export function formatMarkdownReport(target: CodeCoverageColumn, source: CodeCoverageColumn): string {
  const header = `| | ${target.label} | % | ${source.label} | % |
|-:|-:|-:|-:|-:|`;

  const body = target.stat.coverageStats.map(targetStat => {
    const sourceStat = source.stat.coverageStats.find(s => s.label === targetStat.label);
    if (!sourceStat) {
      return '';
    }
    return `| **${targetStat.label}** |
| covered\
 | ${formatNumber(targetStat.covered)}\
 | ${formatPercentage(targetStat.covered, targetStat.total)}\
 | ${formatNumberChange(sourceStat.covered, targetStat.covered, 1)}\
 | ${formatPercentageChange(sourceStat.covered, sourceStat.total, targetStat.covered, targetStat.total, 1)}
| uncovered\
 | ${formatNumber(targetStat.total - targetStat.covered)}\
 | ${formatPercentage(targetStat.total - targetStat.covered, targetStat.total)}\
 | ${formatNumberChange(sourceStat.total - sourceStat.covered, targetStat.total - targetStat.covered, -1)}\
 | ${formatPercentageChange(sourceStat.total - sourceStat.covered, sourceStat.total, targetStat.total - targetStat.covered, targetStat.total, -1)}
| ***total***\
 | ***${formatNumber(targetStat.total)}***\
 |\
 | ${formatNumberChange(sourceStat.total, targetStat.total, 0, '***')} |`;
  }).join('\n');

  return header + '\n' + body;
}