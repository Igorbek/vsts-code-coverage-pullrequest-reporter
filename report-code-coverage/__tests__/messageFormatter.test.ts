import { formatMarkdownReport } from '../messageFormatter';

describe('formatMarkdownReport', () => {
    it('renders markdown message', ( ) => {
        const inputTarget = {
            label: 'master',
            stat: {
                buildFlavor: 'a',
                buildPlatform: 'b',
                coverageStats: [
                    { position: 0, label: 'Lines', covered: 123, total: 234, isDeltaAvailable: false, delta: 0 },
                    { position: 1, label: 'Branches', covered: 234, total: 345, isDeltaAvailable: false, delta: 0 }
                ]
            }
        };
        const inputSource = {
            label: 'PR',
            stat: {
                buildFlavor: 'a',
                buildPlatform: 'b',
                coverageStats: [
                    { position: 0, label: 'Branches', covered: 334, total: 455, isDeltaAvailable: false, delta: 0 },
                    { position: 1, label: 'Lines', covered: 120, total: 235, isDeltaAvailable: false, delta: 0 }
                ]
            }
        };
        const result = formatMarkdownReport(inputTarget, inputSource);

        expect({ inputTarget, inputSource, result }).toMatchSnapshot();
    })
})