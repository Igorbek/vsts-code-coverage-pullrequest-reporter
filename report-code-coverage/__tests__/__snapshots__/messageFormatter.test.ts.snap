// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatMarkdownReport renders markdown message 1`] = `
Object {
  "inputSource": Object {
    "label": "PR",
    "stat": Object {
      "buildFlavor": "a",
      "buildPlatform": "b",
      "coverageStats": Array [
        Object {
          "covered": 334,
          "delta": 0,
          "isDeltaAvailable": false,
          "label": "Branches",
          "position": 0,
          "total": 455,
        },
        Object {
          "covered": 120,
          "delta": 0,
          "isDeltaAvailable": false,
          "label": "Lines",
          "position": 1,
          "total": 235,
        },
      ],
    },
  },
  "inputTarget": Object {
    "label": "master",
    "stat": Object {
      "buildFlavor": "a",
      "buildPlatform": "b",
      "coverageStats": Array [
        Object {
          "covered": 123,
          "delta": 0,
          "isDeltaAvailable": false,
          "label": "Lines",
          "position": 0,
          "total": 234,
        },
        Object {
          "covered": 234,
          "delta": 0,
          "isDeltaAvailable": false,
          "label": "Branches",
          "position": 1,
          "total": 345,
        },
      ],
    },
  },
  "result": "| | master | % | PR | % |
|-:|-:|-:|-:|-:|
| **Lines** |
| covered | 123 | 52.6% | (-3 :arrow_down_small:) 120 :warning: | (-1.5% :arrow_down_small:) 51.1% :warning:
| uncovered | 111 | 47.4% | (+4 :arrow_up_small:) 115 :warning: | (+1.5% :arrow_up_small:) 48.9% :warning:
| ***total*** | ***234*** | | (+***1*** :arrow_up_small:) ***235*** |
| **Branches** |
| covered | 234 | 67.8% | (+100 :arrow_up_small:) 334 :white_check_mark: | (+5.6% :arrow_up_small:) 73.4% :white_check_mark:
| uncovered | 111 | 32.2% | (+10 :arrow_up_small:) 121 :warning: | (-5.6% :arrow_down_small:) 26.6% :white_check_mark:
| ***total*** | ***345*** | | (+***110*** :arrow_up_small:) ***455*** |",
}
`;
